internal/internal_poller_autoscaler_test.go

```diff
@@ -0,0 +1,248 @@
+		name string
+		args args
+		want int
+	}{
+		{
+			name: "dry run don't change anything",
+			args: args{
+				noTaskPoll:               100,
+				taskPoll:                 0,
+				unrelated:                0,
+				initialPollerCount:       10,
+				minPollerCount:           1,
+				maxPollerCount:           10,
+				targetUtilizationInMilli: 500,
+				cooldownTime:             time.Millisecond * 200,
+				isDryRun:                 true,
+			},
+			want: 10,
+		},
+		{
+			name: "no utilization, scale to min",
+			args: args{
+				noTaskPoll:               100,
+				taskPoll:                 0,
+				unrelated:                0,
+				initialPollerCount:       10,
+				minPollerCount:           1,
+				maxPollerCount:           10,
+				targetUtilizationInMilli: 500,
+				cooldownTime:             time.Millisecond * 200,
+				isDryRun:                 false,
+			},
+			want: 1,
+		},
+		{
+			name: "low utilization, scale down",
+			args: args{
+				noTaskPoll:               75,
+				taskPoll:                 25,
+				unrelated:                0,
+				initialPollerCount:       10,
+				minPollerCount:           1,
+				maxPollerCount:           10,
+				targetUtilizationInMilli: 500,
+				cooldownTime:             time.Millisecond * 200,
+				isDryRun:                 false,
+			},
+			want: 5,
+		},
+		{
+			name: "over utilized, scale up",
+			args: args{
+				noTaskPoll:               0,
+				taskPoll:                 100,
+				unrelated:                0,
+				initialPollerCount:       2,
+				minPollerCount:           1,
+				maxPollerCount:           10,
+				targetUtilizationInMilli: 500,
+				cooldownTime:             time.Millisecond * 200,
+				isDryRun:                 false,
+			},
+			want: 4,
+		},
+		{
+			name: "over utilized, scale up to max",
+			args: args{
+				noTaskPoll:               0,
+				taskPoll:                 100,
+				unrelated:                0,
+				initialPollerCount:       6,
+				minPollerCount:           1,
+				maxPollerCount:           10,
+				targetUtilizationInMilli: 500,
+				cooldownTime:             time.Millisecond * 200,
+				isDryRun:                 false,
+			},
+			want: 10,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			pollerScaler := newPollerScaler(
+				tt.args.initialPollerCount,
+				tt.args.minPollerCount,
+				tt.args.maxPollerCount,
+				tt.args.targetUtilizationInMilli,
+				tt.args.isDryRun,
+				tt.args.cooldownTime,
+				zaptest.NewLogger(t))
+			done := pollerScaler.Start()
+			defer done()
+
+			pollChan := generateRandomPollResults(tt.args.noTaskPoll, tt.args.taskPoll, tt.args.unrelated)
+			wg := &sync.WaitGroup{}
+			for i := 0; i < tt.args.maxPollerCount; i++ {
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					for pollResult := range pollChan {
+						pollerScaler.Acquire(1)
+						pollerScaler.CollectUsage(pollResult)
+						pollerScaler.Release(1)
+					}
+				}()
+			}
+			wg.Wait()
+			time.Sleep(time.Millisecond * 500)
```

Comment:
Instead of sleeping yourself, better would be to use assert.Eventually() - less chance of race condition and should test faster.

Reply:
TIL. Thank you for the advice. I've added optional hook functions in the pollerAutoscaler implementation to simply the testing.